posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+
+Well there's a reason! By making `UserError` a subtype of `Throwable` we can take advantage of `Prisms` by going back and forth in the types. See what I'm going yet?
+
+`UserRoute` has a `Sync[F]` constraint meaning that we have available a `MonadError[F, Throwable]` instance but we would like to have `MonadError[F, UserError]` instead to leverage the Scala compiler. There's a caveat and is that the error types need to be of the same family so we can derive a `Prism` that can navigate the errors types in one direction or another. But how do we derive it?
+
+#### Cats Meow MTL
+
+Fortunately our friend [Oleg Pyzhcov](https://twitter.com/oleg_pyzhcov) has created this great library named [meow-mtl](https://github.com/oleg-py/meow-mtl) that makes heavy use of [Shapeless](https://github.com/milessabin/shapeless) in order to derive `Lenses` and `Prisms` and it provides instances for some `cats-effect` compatible datatypes.
+
+And two of the supported typeclasses are `ApplicativeError` and `MonadError` as long as the error type is a subtype of `Throwable` to make it compatible with `cats-effect`. So we can do something like this:
+
+```scala
+import com.olegpy.meow.hierarchy._ // All you need is this import!
+
+case class CustomError(msg: String) extends Throwable
+
+def customHandle[F[_]: MonadError[?[_], CustomError], A](f: F[A], fallback: F[A]): F[A] =
+  f.handleErrorWith(_ => fallback)
+
+val io: IO[Int] = ???
+customHandle(io, IO.pure(123))
+```
+
+#### Generalizing Http Error Handling
+
+Now back to our use case we can't have a `MonadError[F, UserError]` constraint because there's already a `MonadError[F, Throwable]` in scope given our `Sync[F]` constraint. But it turns out we can make this work if we also abstract over the error handling by introducing an `HttpErrorHandler` algebra where the error type is a subtype of `Throwable`.
+
+```tut:book:silent
+import cats.MonadError
+
+trait HttpErrorHandler[F[_], E <: Throwable] {
+  def handle(fa: F[Response[F]]): F[Response[F]]
+}
+
+object HttpErrorHandler {
+  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
+}
+
+object syntax {
+  implicit class HttpErrorHandlerOps[F[_], E <: Throwable](fa: F[Response[F]]) {
+    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): F[Response[F]] =
+      H.handle(fa)
+  }
+}
+```
+
+`UserRoutes` can now have an additional constraint of type `HttpErrorHandler[F, UserError]` so we clearly know what kind of errors we are dealing with and can have the Scala compiler on our side.
+
+```tut:book:silent
+class UserRoutesMTL[F[_]: Sync](userAlgebra: UserAlgebra[F])(implicit ev: HttpErrorHandler[F, UserError]) extends Http4sDsl[F] {
+
+  import JsonCodecs._
+  import syntax._
+
+  val routes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      userAlgebra.find(username).flatMap {
+        case Some(user) => Ok(user)
+        case None => NotFound(username)
+      }.handleHttpErrorResponse
```

Comment:
This gets repeated for every route.  Can we go one step further and put `handleHttpErrorResponse` on `routes` itself?  It compiles in my head...

Reply:
Sounds great! I'll try to get it compiling and working :)

Reply:
@rossabaker can you expand on this idea? I'm now trying something out but nothing seems correct. This is what I've come up with so far (compiling):

```scala
trait HttpErrorHandler[F[_], E <: Throwable] {
  def handle(fa: OptionT[F, Response[F]]): OptionT[F, Response[F]]
}

object HttpErrorHandler {
  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
}

object syntax {
  implicit class HttpErrorHandlerOp[F[_]: Functor, E <: Throwable](routes: HttpRoutes[F]) {
    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): HttpRoutes[F] = {
//      import org.http4s.implicits._
//      val app: HttpApp[F] = routes.orNotFound // adding orNotFound here does not seem correct
//      app.mapF(fa => OptionT.liftF(H.handle(fa)))
        routes.mapF(H.handleOpt)
    }
  }
}

class UserRoutesMTL[F[_]: Sync](userAlgebra: UserAlgebra[F])
                               (implicit H: HttpErrorHandler[F, UserError]) extends Http4sDsl[F] {

  private val httpRoutes: HttpRoutes[F] = HttpRoutes.of[F] { ... }

  val routes: HttpRoutes[F] = httpRoutes.handleHttpErrorResponse
}

class UserHttpErrorHandler[F[_]: MonadError[?[_], UserError]] extends HttpErrorHandler[F, UserError] with Http4sDsl[F] {
  override def handleOpt(fa: OptionT[F, Response[F]]) = {
    OptionT {
      fa.value.handleErrorWith {
        case InvalidUserAge(age) => BadRequest(s"Invalid age $age".asJson).map(Option(_)) // argh
        case UserAlreadyExists(username) => Conflict(username.asJson).map(Option(_))
        case UserNotFound(username) => NotFound(username.asJson).map(Option(_))
      }
    }
  }
}
```

Reply:
Well this one seems a bit more acceptable but I'd like to hear how you would do it :smile: 

```scala

trait HttpErrorHandler[F[_], E <: Throwable] {
 def handle(routes: HttpRoutes[F]): HttpRoutes[F]
}

object HttpErrorHandler {
  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
}

object syntax {
  implicit class HttpErrorHandlerOp[F[_]: Functor, E <: Throwable](routes: HttpRoutes[F]) {
    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): HttpRoutes[F] = {
      H.handle(routes)
    }
  }
}

class UserHttpErrorHandler[F[_]: MonadError[?[_], UserError]] extends HttpErrorHandler[F, UserError] with Http4sDsl[F] {
  override def handle(routes: HttpRoutes[F]): HttpRoutes[F] = {
    val handler: UserError => F[Response[F]] = {
      case InvalidUserAge(age) => BadRequest(s"Invalid age $age".asJson)
      case UserAlreadyExists(username) => Conflict(username.asJson)
      case UserNotFound(username) => NotFound(username.asJson)
    }
    
    Kleisli { req =>
      OptionT {
        routes.run(req).value.handleErrorWith(e => handler(e).map(Option(_)))
      }
    }
  }
}
```

Reply:
And going one step further I can extract that Kleisli stuff out in an abstract class:

```scala
trait HttpErrorHandler[F[_], E <: Throwable] {
  def handle: E => F[Response[F]]
  def routesHandler: HttpRoutes[F] => HttpRoutes[F]
}

abstract class DefaultHttpErrorHandler[F[_]: MonadError[?[_], E], E <: Throwable] extends HttpErrorHandler[F, E] {
  override def routesHandler: HttpRoutes[F] => HttpRoutes[F] = { routes =>
    Kleisli { req =>
      OptionT {
        routes.run(req).value.handleErrorWith(e => handle(e).map(Option(_)))
      }
    }
  }
}

object HttpErrorHandler {
  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
}

object syntax {
  implicit class HttpErrorHandlerOps[F[_], E <: Throwable](routes: HttpRoutes[F]) {
    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): HttpRoutes[F] = {
      H.routesHandler(routes)
    }
  }
}

class UserHttpErrorHandler[F[_]: MonadError[?[_], UserError]] extends DefaultHttpErrorHandler[F, UserError] with Http4sDsl[F] {
  override def handle: UserError => F[Response[F]] = {
    case InvalidUserAge(age) => BadRequest(s"Invalid age $age".asJson)
    case UserAlreadyExists(username) => Conflict(username.asJson)
    case UserNotFound(username) => NotFound(username.asJson)
  }
}
```

Looking forward to your feedback!!! :smile:

Reply:
And the last one (woke up with some clearer ideas :smile: )

```scala
trait HttpErrorHandler[F[_], E <: Throwable] {
  def handle(routes: HttpRoutes[F]): HttpRoutes[F]
}

object RoutesHttpErrorHandler {
  def apply[F[_]: MonadError[?[_], E], E <: Throwable](routes: HttpRoutes[F])(handler: E => F[Response[F]]): HttpRoutes[F] = {
    Kleisli { req =>
      OptionT {
        routes.run(req).value.handleErrorWith(e => handler(e).map(Option(_)))
      }
    }
  }
}

object HttpErrorHandler {
  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
}

object syntax {
  implicit class HttpErrorHandlerOp[F[_], E <: Throwable](routes: HttpRoutes[F]) {
    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): HttpRoutes[F] = {
      H.handle(routes)
    }
  }
}

class UserHttpErrorHandler[F[_]: MonadError[?[_], UserError]] extends HttpErrorHandler[F, UserError] with Http4sDsl[F] {
  private val handler: UserError => F[Response[F]] = {
    case InvalidUserAge(age) => BadRequest(s"Invalid age $age".asJson)
    case UserAlreadyExists(username) => Conflict(username.asJson)
    case UserNotFound(username) => NotFound(username.asJson)
  }

  override def handle(routes: HttpRoutes[F]): HttpRoutes[F] =
    RoutesHttpErrorHandler(routes)(handler)
}
```